### PR TITLE
Add workflow matrices for the tests to reduce GitHub check runtime

### PR DESCRIPTION
**Pull request by nabenabe0928**
## Motivation

As the CI is slow with tests on all the Python versions, I made changes so that the tests will run only on the oldest and the latest supported Python versions.
For the first step, I applied the similar change to tests mpi in [PR#5067](https://github.com/optuna/optuna/pull/5067), which is not one of the required tests.
In this PR, I further change the other tests to reduce the runtime.
Note that these changes require modifying the requirements of CI.

## Description of the changes

More specifically, the changes are:
1. Remove the Python versions 3.9, 3.10 from tests-integration,
2. Remove the Python versions 3.8, 3.9, 3.10 from tests-storage, and
3. Remove the Python versions 3.8, 3.9, 3.10, 3.11 from tests.

### CODE REVIEW

**Comment by HideakiImamura** [comment]
@Alnusjaponica @not522 Could you review this PR?

File: .github/workflows/tests-mpi.yml [hunk 1]
```diff
@@ -90,10 +90,7 @@ jobs:
     - name: Tests
       run: |
         export OMPI_MCA_rmaps_base_oversubscribe=yes
-
-        if [ ${{ matrix.python-version }} != 3.11 ]; then
-          mpirun -n 2 -- pytest tests/integration_tests/test_pytorch_distributed.py
-        fi
+        mpirun -n 2 -- pytest tests/integration_tests/test_pytorch_distributed.py
```
Comment (not522) [inline .github/workflows/tests-mpi.yml:93]: Could you separate this change to another PR?

Comment (nabenabe0928) [reply .github/workflows/tests-mpi.yml]: Done!

**Review by Alnusjaponica — COMMENTED** [review]
There is a more simple work around (see https://github.com/orgs/community/discussions/26253#discussioncomment-3250989) and it actually works (see https://github.com/Alnusjaponica/optuna/pull/25). Any thoughts on this?

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -14,13 +14,30 @@ concurrency:
...
+    steps:
+    - name: Define workflow matrix
+      id: set-matrix
+      run: |
+        if [ "${{ github.event_name }}" = "schedule" ] || [ "${{ github.event_name }}" = "workflow_dispatch" ] ; then
+          echo "Schedule tests (run all the versions)"
+          echo "matrix={\"python-version\":[\"3.7\", \"3.8\", \"3.9\", \"3.10\", \"3.11\", \"3.12\"]}" >> $GITHUB_OUTPUT
+        else
+          echo "PR push tests (run only the oldest and the latest versions)"
+          # QUESTION: Shall I run 3.11 as well because BoTorch will not be tested on 3.12 right now?
```
[the remark this answers is not in the document]
Comment (Alnusjaponica) [reply .github/workflows/tests.yml:30]: Another possible workaround is here, though I haven't confirmed if it works.
https://github.com/orgs/community/discussions/9044#discussion-3754049

**Comment by Alnusjaponica** [comment]
Another possible workaround is here, though I haven't confirmed if it works.
https://github.com/orgs/community/discussions/9044#discussion-3754049

**Comment by nabenabe0928** [comment]
> There is a more simple work around (see https://github.com/orgs/community/discussions/26253#discussioncomment-3250989) and it actually works (see [Alnusjaponica#25](https://github.com/Alnusjaponica/optuna/pull/25)). Any thoughts on this?

I applied your request to `tests-mpi.yml`, but have not applied to the other files yet.
As we already agreed on the current version, I would like to discuss the change in `tests-mpi` with our team and get back to you again.

**Comment by nabenabe0928** [comment]
> > There is a more simple work around (see https://github.com/orgs/community/discussions/26253#discussioncomment-3250989) and it actually works (see [Alnusjaponica#25](https://github.com/Alnusjaponica/optuna/pull/25)). Any thoughts on this?
> 
> I applied your request to `tests-mpi.yml`, but have not applied to the other files yet. As we already agreed on the current version, I would like to discuss the change in `tests-mpi` with our team and get back to you again.

Also, I picked up a comment from ChatGPT4 that says:

_"The behavior of complex expressions within the matrix construct is not explicitly documented by GitHub and our workflow might suddenly stop working (`GITHUB_OUTPUT` is explicitly documented)."_ 

I am not sure how likely it will happen, but just in case, this change potentially embeds a future bug.

**Review by not522 — COMMENTED** [review]
Could you add `workflow_dispatch` for trigger events of `tests-storage.yml`?

**Comment by nabenabe0928** [comment]
> Could you add `workflow_dispatch` for trigger events of `tests-storage.yml`?

@not522 Done:)
Please take another look!

**Review by not522 — APPROVED** [review]
LGTM!

**Review by Alnusjaponica — APPROVED** [review]
LGTM. 
@not522 Could you merge this PR?

**Comment by not522** [comment]
The CI names have extra true/false. How about doing like this? https://github.com/not522/optuna/commit/1addaabd1f9987a2df84416508fc2dabfe3e597b

**Comment by nabenabe0928** [comment]
> The CI names have extra true/false. How about doing like this? [not522@1addaab](https://github.com/not522/optuna/commit/1addaabd1f9987a2df84416508fc2dabfe3e597b)

I reflected your suggestion on the PR with a slight modification.
Basically, I changed so that "PR" will be shown for each push to a PR and "Scheduled" will be shown otherwise.

**Review by not522 — APPROVED** [review]
Thank you for your update! LGTM!
